### PR TITLE
Add fs iotime in Nodes Stats API

### PR DESCRIPTION
**Pull request by easyice**
Related to #67805

This PR add io_time_in_millis in Nodes stats API,  In the benchmark, it works almost as well as the iostat command

step 1,Perform random reads and writes on disk

````
fio -filename=/data00/tmp/1Gb.file -direct=1 -iodepth 1 -thread -rw=randrw -ioengine=libaio -bs=4k -size=10G -numjobs=2 -runtime=1000 -group_reporting -name=mytest
````

step 2, Compare the results of the io_time and iostat commands,use this script

````
while true
do
 ts1=`echo $(($(date +%s%N)/1000000))`
 iotime1=`curl -s "127.0.0.1:9200/_nodes/stats/fs?pretty" |grep io_time_in_millis | head -n 1 | awk '{print $3}'`
 sleep 5
 iotime2=`curl -s "127.0.0.1:9200/_nodes/stats/fs?pretty" |grep io_time_in_millis | head -n 1 | awk '{print $3}'`
 ts2=`echo $(($(date +%s%N)/1000000))`

 echo "scale=2;($iotime2-$iotime1)/($ts2-$ts1)*100" | bc
done
````

the result:

image

### CODE REVIEW

**Comment by mark-vieira** [comment]
@elasticmachine update branch

**Review by andreidan — APPROVED** [review]
Thanks for working on this @easyice. LGTM

**Comment by andreidan** [comment]
@elasticmachine ok to test

**Comment by andreidan** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java [hunk 1]
```diff
@@ -242,6 +253,8 @@ public void writeTo(StreamOutput out) throws IOException {
             out.writeLong(previousSectorsRead);
             out.writeLong(currentSectorsWritten);
             out.writeLong(previousSectorsWritten);
+            out.writeLong(currentIOTime);
+            out.writeLong(previousIOTime);
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java:257]: We'll need to read/write the new fields in a backwards compatible way (similar to how we do it here for eg. https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/datafeed/AggProvider.java#L137 - there are probably other areas in the code). Let me know if you need any help with these version checks to ensure these fields are added in a BWC way.

File: server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java [hunk 2]
```diff
@@ -227,6 +236,8 @@ public DeviceStats(StreamInput in) throws IOException {
             previousSectorsRead = in.readLong();
             currentSectorsWritten = in.readLong();
             previousSectorsWritten = in.readLong();
+            currentIOTime = in.readLong();
+            previousIOTime = in.readLong();
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java:240]: Version checks for bwc are needed here as well

**Review by andreidan — CHANGES_REQUESTED** [review]
@easyice I've just left a comment regarding adding backwards compatibility checks for this change. Thanks!

File: server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java (hunk 1 above)
Comment (easyice) [reply server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java]: Thanks for the detailed instructions, i added  the compatibility checks ,use Version.V_7_13_0

File: server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java (hunk 2 above)
Comment (easyice) [reply server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java]: Thanks for review,  Version checks is added

**Review by andreidan — CHANGES_REQUESTED** [review]
Thanks for iterating on this @easyice 

Apologies, I should've mentioned, at the time of merge we need the version checks to be against `V8_0_0` (and I'll subsequently change them to `7_13_0` after the change is backported to that release line). Can you please modify the version checks? 

Also, there are some checkstyles issues. We have more information on how to configure and run the checkstyle locally [here](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#checkstyle) - alternatively, running something along the lines of `./gradlew checkStyleMain checkStyleTest checkstyleInternalClusterTest checkstyleIntegTest checkstyleTestFixtures` should help spot the issues

**Comment by easyice** [comment]
Thanks for @andreidan , I will pay more attention next time

File: server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java [hunk 3]
```diff
@@ -166,6 +167,8 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         final long previousWritesCompleted;
         final long currentSectorsWritten;
         final long previousSectorsWritten;
+        long currentIOTime;
+        long previousIOTime;
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java:171]: Can we also mark these as `final`?

**Review by andreidan — CHANGES_REQUESTED** [review]
Thanks for fixing the version and the checkstyle

Added just a few more minor comments, this is close to complete.

Besides the comments there's also a test failure in `org.elasticsearch.xpack.monitoring.collector.node.NodeStatsMonitoringDocTests.testToXContent`

File: server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java [hunk 4]
```diff
@@ -295,13 +320,16 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
+        private static final String IO_TIMEMS = "io_time_in_millis";
+
 
         final DeviceStats[] devicesStats;
         final long totalOperations;
         final long totalReadOperations;
         final long totalWriteOperations;
         final long totalReadKilobytes;
         final long totalWriteKilobytes;
+        long totalIOTimeInMillis;
```
[the remark this answers is not in the document]
Comment (easyice) [reply server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java]: thanks you  @andreidan, i will make  it as final

File: server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java (hunk 3 above)
Comment (easyice) [reply server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java]: Okay, I've changed it to final, thank you for check!

**Comment by easyice** [comment]
Thanks for @andreidan 
The test failure  for NodeStatsMonitoringDocTests.testToXContent is fixed, but the io_stats field is not add in testToXContent function:  

https://github.com/elastic/elasticsearch/blob/f0aa1b1f1f8470659fcf98e96ece9e959f7a6844/x-pack/plugin/monitoring/src/test/java/org/elasticsearch/xpack/monitoring/collector/node/NodeStatsMonitoringDocTests.java#L277-L283

and NodeStatsMonitoringDoc#XCONTENT_FILTERS
https://github.com/elastic/elasticsearch/blob/f0aa1b1f1f8470659fcf98e96ece9e959f7a6844/x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/collector/node/NodeStatsMonitoringDoc.java#L115-L120

Do these also need to be add the `io_time_in_millis` field?

**Comment by easyice** [comment]
I'm sorry i use another computer push some code, which git config is not match this user and email, cause the cla check error, when I change to the correct configuration, the cla check still returns an error, can you please help me how to fix this issue? @andreidan

**Comment by easyice** [comment]
@andreidan  the CLA check failed  issue is fixed, Can you please take review again? Thanks!

File: server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java [hunk 5]
```diff
@@ -336,6 +370,12 @@ public IoStats(StreamInput in) throws IOException {
             this.totalWriteOperations = in.readLong();
             this.totalReadKilobytes = in.readLong();
             this.totalWriteKilobytes = in.readLong();
+            if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
+                this.totalIOTimeInMillis = in.readLong();
+            }
+            else{
+                this.totalIOTimeInMillis = 0;
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java:377]: super minor: there's a formatting issue here

File: server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java [hunk 6]
```diff
@@ -227,6 +237,13 @@ public DeviceStats(StreamInput in) throws IOException {
             previousSectorsRead = in.readLong();
             currentSectorsWritten = in.readLong();
             previousSectorsWritten = in.readLong();
+            if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
+                currentIOTime = in.readLong();
+                previousIOTime = in.readLong();
+            }else {
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/monitor/fs/FsInfo.java:243]: super minor: there's a formatting issue here

**Review by andreidan — COMMENTED** [review]
Sorry for the late reply @easyice (I was on holiday). Thanks for addressing the previous comments and for iterating on this! I think this is close to complete, I've left just a couple of super minor formatting comments after which this is complete

**Comment by andreidan** [comment]
@elasticmachine update branch

**Comment by easyice** [comment]
Thanks for @andreidan , happy holiday!  the formatting issue is fixed

**Comment by andreidan** [comment]
@elasticmachine update branch

**Review by andreidan — APPROVED** [review]
Thanks for your contribution and for iterating on this @easyice 

LGTM

**Comment by easyice** [comment]
Thanks for your help @andreidan

**Pull request merged**